invoke/completion/complete.py

```diff
@@ -25,13 +37,16 @@ def complete(names, core, initial_context, collection, parser):
         # Gently parse invocation to obtain 'current' context.
         # Use last seen context in case of failure (required for
         # otherwise-invalid partial invocations being completed).
+
+        # FIXME: this seems wonky
+        contexts: Union[List[ParserContext], ParseResult]
```

Comment:
Not sure if there about how this one could be improved.

Reply:
ParseResult appears to subclass list. So, I think it should be tweaked to subclass `List[ParserContext]`, or whatever is needed to make it work at runtime. Then this annotation will work with just `List[ParserContext]`.

Reply:
My understanding after a brief search is that the actual "right" way to do this is to change `ParseResult` to inherit from [MutableSequence](https://docs.python.org/3/library/collections.abc.html#collections.abc.MutableSequence) instead of from `list`, since "a listy thing that may only hold ParserContext objects can't be substituted for any other list".

Is it actually possible to just say "subclasses `List[ParserContext]`"?

For the time being, it's _probably_ best to leave this line (and the class inheritance) alone since changing it is probably technically backwards incompatible, but that's a very weak assertion on my part -- so I'm open to counterarguments if this "contexts is xyz type" line is incredibly gross somehow.

Reply:
> My understanding after a brief search is that the actual "right" way to do this is to change `ParseResult` to inherit from [MutableSequence](https://docs.python.org/3/library/collections.abc.html#collections.abc.MutableSequence) instead of from `list`, since "a listy thing that may only hold ParserContext objects can't be substituted for any other list".

`MutableSequence` is an abc. So, if you implement all of the methods for that abc, you can do that. But, subclassing from `list` is also perfectly valid and possibly beneficial if it gives you several methods for free.

From a typing perspective `list[ParserContext]` is a list that can't be substituted for other lists which might contain other types. In this situation, it doesn't look like MutableSequence is needed to me.

> Is it actually possible to just say "subclasses `List[ParserContext]`"?

Generally an annotation refers to subclasses, so unless I'm misunderstanding your comment, that already means what you want it to mean.

> For the time being, it's _probably_ best to leave this line (and the class inheritance) alone since changing it is technically backwards incompatible (i.e. clean this up next major rel), but that's a very weak assertion on my part -- I'm open to counterarguments if this "contexts is xyz type" line is incredibly gross somehow.

If we were supporting Python 3.9+ only, then we would subclass `list[ParserContext]`, which at runtime is identical to subclassing `list`. To support Python 3.7/8, we need to make use of `typing.List` though. I just tested it at runtime, and subclassing `List` works fine, and `isinstance(custom_list, list)` passes successfully too. So, I think this change is safe and backwards compatible.

Reply:
Yeah, I think I might need to clarify. I already knew `ParseResult` was a list, but kept forgetting about it when I was looking at it. TBC, this seems more like a issue communicating intent.

To me it reads as it should be `List[Union[ParserContext, ParseResult]]` or `Union[List[ParserContext], Parseresults]`.

Hope that helps.

Reply:
Is this resolved?

Reply:
It is now, look out for an upcoming push from me. tl;dr Sam was right that we can just subclass `List[ParserContext]` instead of `list`. This simplifies the referenced spot so it no longer requires an awkward Union; it also found a minor typing mistake elsewhere. Hey, the system works!